ZydisEncoder/generate_encoder_types.py

```diff
@@ -0,0 +1,255 @@
+            elif filename == 'ZydisExportConfig.h':
+                return open(os.path.join(self.zydis_path, 'assets', 'ZydisExportConfigSample.h'))
+
+        return super().on_file_open(is_system_include, includepath)
+
+    def get_preprocessed_file(self):
+        file = io.StringIO()
+        self.write(file)
+        file.seek(0)
+        return file.read()
+
+    @staticmethod
+    def concatenate_tokens(tokens):
+        return ''.join([token.value for token in tokens])
+
+    def expand_macro(self, macro):
+        if isinstance(macro, str):
+            macro = self.macros[macro].value
+        else:
+            macro = macro.value
+        return self.concatenate_tokens(self.expand_macros(macro))
+
+
+class ZydisParser(object):
+
+    class ZydisParsingError(Exception):
+        pass
+
+    class EnumVisitor(c_ast.NodeVisitor):
+
+        def __init__(self, typenames):
+            self.typenames = typenames
+            self.parsed_types = {}
+
+        @staticmethod
+        def get_const_value(node):
+            if node is None:
+                return None
+            if node.type != 'int':
+                raise ZydisParser.ZydisParsingError('Got enum value of type %s (expected int)' % node.type)
+            if node.value.startswith('0x'):
+                return int(node.value, 16)
+            else:
+                return int(node.value)
+
+        def visit_Typedef(self, node):
+            if node.name in self.typenames:
+                self.parsed_types[node.name] = [(e.name, self.get_const_value(e.value)) for e in node.type.type.values.enumerators
+                                                if not e.name.endswith('_MAX_VALUE') and not e.name.endswith('_REQUIRED_BITS')]
+
+    def __init__(self, file, filename):
+        parser = c_parser.CParser()
+        self.ast = parser.parse(file, filename=filename)
+
+    def get_enum_types(self, typenames):
+        v = ZydisParser.EnumVisitor(typenames)
+        v.visit(self.ast)
+        types = ''
+        for typename in typenames:
+            if typename not in v.parsed_types:
+                raise ZydisParser.ZydisParsingError('Type %s not found' % typename)
+            enumerators = v.parsed_types[typename]
+            value_type = type(enumerators[0][1])
+            if not all([isinstance(e[1], value_type) for e in enumerators]):
+                raise ZydisParser.ZydisParsingError('%s: All values must be explicitly assigned or no values must be assigned' % typename)
+
+            is_flag_enum = False
+            if value_type == int:
+                popcnts = [bin(e[1]).count('1') == 1 for e in enumerators]
+                is_flag_enum = not popcnts[0] and all(popcnts[1:]) and \
+                               [e[1] for e in enumerators][1:] == [2**n for n in range(len(enumerators) - 1)]
+            if is_flag_enum:
+                types += "%s = IntFlag('%s', [\n" % (typename, typename)
```

Comment:
Just a heads-up that there are format string literals in Python 3 that allow you to write:

```suggestion
                types += f"{typename} = IntFlag('{typename}', [\n"
```

However, since the whole Python stuff is rather temporary and for the sake of consistency, feel free to keep as-is, I don't really mind.

Reply:
I know about them but I'm guilty of being a hardcore printf-style user. It's just a personal preference. That being said there are cases like this one where modern formatting approaches are clearly superior. Also they are 3.6+ and that used to be a deal-breaker on distributions that shipped with older Python, so I had a fairly good excuse to avoid them :P

I would like to keep it as-is for consistency with all other places (because I really don't want to change them all).

Reply:
Yeah, that's fine!